manager/orchestrator/constraintenforcer/constraint_enforcer.go

```diff
@@ -115,16 +122,27 @@ func (ce *ConstraintEnforcer) rejectNoncompliantTasks(node *api.Node) {
 		// Ensure that the task assigned to the node
 		// still satisfies the resource limits.
 		if t.Spec.Resources != nil && t.Spec.Resources.Reservations != nil {
-			if t.Spec.Resources.Reservations.MemoryBytes > availableMemoryBytes {
+			if t.Spec.Resources.Reservations.MemoryBytes > available.MemoryBytes {
 				removeTasks[t.ID] = t
 				continue
 			}
-			if t.Spec.Resources.Reservations.NanoCPUs > availableNanoCPUs {
+			if t.Spec.Resources.Reservations.NanoCPUs > available.NanoCPUs {
 				removeTasks[t.ID] = t
 				continue
 			}
-			availableMemoryBytes -= t.Spec.Resources.Reservations.MemoryBytes
-			availableNanoCPUs -= t.Spec.Resources.Reservations.NanoCPUs
+			for _, ta := range t.AssignedGenericResources {
+				// Type change or no longer available
+				if genericresource.HasResource(ta, available.Generic) {
+					removeTasks[t.ID] = t
+					break loop
+				}
+			}
+
+			available.MemoryBytes -= t.Spec.Resources.Reservations.MemoryBytes
+			available.NanoCPUs -= t.Spec.Resources.Reservations.NanoCPUs
+
+			genericresource.ClaimResources(&available.Generic,
+				&fakeStore, t.AssignedGenericResources)
```

Comment:
> However, I think we were discussing this at some point and I was under the impression that it was not the job of the scheduler (through "reconciliation") to check if the resources were still available for the jobs.

This (and the rest of your comment) is correct.

My thinking was that the jobs are similar enough that we might be able to use the same function to do both things. Both the scheduler and constraint enforcer could call `RemoveNodeResources` to subtract the available resources used by each task. If there were insufficient resources, `RemoveNodeResources` would return an error. The constraint enforcer would act on this error by shutting down the task. The scheduler would ignore the error, since it is just updating its accounting.

Now that I think about it, though, we probably do want different behavior between the two. If the scheduler encounters a situation where too much of a resource is used, it should let the value go negative so that it ends up correct after `removeTask` when the task is shut down (right?). With the constraint enforcer, values should never go negative. Does this make sense?